### ISSUE TITLE
Alt tag for inserted images

### ISSUE THREAD

**Issue by acapro** [denkGroot/Spina#237]
I would like to have functionality for adding alt tags to inserted images (from the trix editor).

I noticed when adding an alt tag to `this.editor.insertHTML "<img src='#{ url }' alt='some desc' />"` the trix document model/parser ignores it altogether, I posted an issue [here](https://github.com/basecamp/trix/issues/384)

Any ideas? I'm pretty much stuck at this.

Also I noticed you added the following to trix.coffee.erb on line 37 to 38 (inside the`@photoInsert` method):

    length = this.editor.getDocument().toString().length
    this.editor.setSelectedRange(length - 1)

This causes to always add the photo to the bottom, any special reason for this? Removing this allows photos to be inserted where the cursor is, which in my opinion is the expected functionality.

**Comment by danmitchell-** [comment]
If you want more control than Trix provides then id recommend swapping it for this gem https://github.com/COSMOSMagazine/spina_summernote

**Comment by acapro** [comment]
Thanks! Tried it out, looks promising (almost too much control :) ) Would really like to stick with trix however.  I noticed that it does allow for a figcaption, not sure if it is enough to replace an alt tag - hard to find out. What i did find was from [here](http://davidmacd.com/blog/figcaption.html)

> HTML5 does *not* allow figcaption as an alternative to alt

I would think this could be an important feature, since trix being used in spina to modify content of actual webpages and not just a comment for example.  So for accessibility and seo pretty much a must have.

**Comment by danmitchell-** [comment]
You can easily remove control with the summer note plugin but I take your point on sticking with Trix and making these SEO improvements. I's love to see '<p>' instead of '<div>' too :trollface:

**Comment by Bramjetten** [comment]
In my opinion, the best way forward is to extend Trix. I think the way Trix works is far superior to the vast majority of so-called WYSIWYG-editors. I've tried a LOT of them and I can say that Trix is by far the most reliable.

**Comment by danmitchell-** [comment]
Agreed. Prehaps a list features for the roadmap would be the way forward then. 

These are requirements I'd personally like to see:
- Be able to add alt tags to inserted images
- Be able to add a link to an image
- Be able to insert an email link as well as a page link
- Be able to link to a `Spina::Page` as well as an absolute or relative URL (incase the slug changes)
- Render basic content as `<p>` and not `<div>`

**Comment by Bramjetten** [comment]
Sounds great @danmitchell- 
The only thing I'm not convinced about is switching to <p>-tags. I mean sure, when writing paragraphs of text they are semantically correct, but I cannot find evidence it has any meaningful impact on SEO. <div>-tags are easier to work with from a developer's point of view.

**Comment by acapro** [comment]
there is a discussion about using the `<p>`tag [here](https://github.com/basecamp/trix/issues/202)

**Comment by acapro** [comment]
I've been digging into the trix source code, and figured out how to extend the insertHTML method with an alt tag. First I copied the src folder from the trix repository to the vendor folder, then in `spina/admin/application.coffee.erb`replaced the trix inclusion with `#= require trix/index`

I works immediately, witch is awesome - so I can start editing the source code.  What I did was the following:

trix/models/html_parser.coffee | line 5 | added alt to `allowedAttributes`:
```
allowedAttributes = "style href src width height class alt".split(" ")
```

trix/model/attachment.coffee | line 66 | added a `getAlt` method:
```
getURL: ->
  @attributes.get("url")
getHref: ->
  @attributes.get("href")
getAlt: ->
  @attributes.get("alt")
```

trix/views/previewable_attachment_view.coffee | line 10 | add alt attribute to createContentNodes:
```
createContentNodes: ->
  @image = makeElement
    tagName: "img"
      attributes:
        src: ""
        alt: ""
      data:
        trixMutable: true
```

and in the same file appended the updateAttributesForImage method:
```
updateAttributesForImage: (image) ->
  url = @attachment.getURL()
  previewURL = @attachment.getPreviewURL()
  image.src = previewURL or url
  image.alt = @attachment.getAlt()
```

now when calling the insertHTML method like so:
```
this.editor.insertHTML "<img src='#{ url }' alt='test' />"
```
the image will have an alt tag 👍 
hope this helps

**Comment by danmitchell-** [comment]
A PR would be great @acapro

**Comment by acapro** [comment]
done!

**Comment by acapro** [comment]
![photo-options](https://cloud.githubusercontent.com/assets/4762057/24401962/d4c6ceba-13b6-11e7-8a9f-284e4ade9452.gif)

**Comment by danmitchell-** [comment]
Sweet! Looks good. We can probably add a link field in that modal too
